sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,473 @@
+
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An unbounded source for MQTT broker.
+ *
+ * <h3>Reading from a MQTT broker</h3>
+ *
+ * <p>MqttIO source returns an unbounded collection of {@code byte[]} as
+ * {@code PCollection<byte[]>}, where {@code byte[]} is the message payload.</p>
+ *
+ * <p>To configure a MQTT source, you have to provide a {@code ClientId}, a {@code ServerURI},
+ * and eventually a {@code Topic} pattern. The following example illustrates various options
+ * for configuring the source:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(MqttIO.read().withClientId("my_client").withServerUri("tcp://host:11883")
+ *   .withTopic("topic")
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a MQTT broker</h3>
+ *
+ * <p>MqttIO sink supports writing {@code byte[]} to a topic on a MQTT broker.</p>
+ *
+ * <p>To configure a MQTT sink, you must specify the {@code ClientId}, {@code ServerURI},
+ * {@code Topic}. Eventually, you can also specify the {@code Retained} and {@code QoS} of the
+ * message.</p>
+ *
+ * <p>For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withClientId("my-client")
+ *     .withServerUri("tcp://host:11883")
+ *     .withTopic("my-topic")
+ *     .withRetained(true)
+ *     .withQoS(2)
+ *
+ * }</pre>
+ */
+public class MqttIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MqttIO.class);
+
+  public static Read read() {
+    return new Read(new UnboundedMqttSource(null, null, null,
+        Long.MAX_VALUE, null));
+  }
+
+  public static Write write() {
+    return new Write(new Write.MqttWriter(null, null, null, 1, false));
+  }
+
+  private MqttIO() {
+  }
+
+  /**
+   * A {@link PTransform<PBegin, PCollection<byte[]>>} to read from a MQTT broker.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<byte[]>> {
+
+    public Read withServerUri(String serverUri) {
+      return new Read(source.withServerUri(serverUri));
+    }
+
+    public Read withClientId(String clientId) {
+      return new Read(source.withClientId(clientId));
+    }
+
+    public Read withTopic(String topic) {
+      return new Read(source.withTopic(topic));
+    }
+
+    public Read withMaxNumRecords(long maxNumRecords) {
+      return new Read(source.withMaxNumRecords(maxNumRecords));
+    }
+
+    public Read withMaxReadTime(Duration maxReadTime) {
+      return new Read(source.withMaxReadTime(maxReadTime));
+    }
+
+    private final UnboundedMqttSource source;
+
+    private Read(UnboundedMqttSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public PCollection<byte[]> apply(PBegin input) {
+
+      org.apache.beam.sdk.io.Read.Unbounded<byte[]> unbounded =
+          org.apache.beam.sdk.io.Read.from(getSource());
+
+      PTransform<PBegin, PCollection<byte[]>> transform = unbounded;
```

Comment:
This snippet is repeating in all unbounded sources we have. What do you think about deduplicating it? E.g. making .withMaxNumRecords() special-case Long.MAX_VALUE, and making withMaxReadTime special-case Duration.millis(Long.MAX_VALUE) to do exactly this, so that you can say:

return input.getPipeline().apply(Read.from(getSource()).withMaxNumRecords(maxNumRecords).withMaxReadTime(maxReadTime));

And one more thing: maxNumRecords and maxReadTime are not part of the source - they're not used in the code of the source; they're actually configuration of the transform *on top of* the source.